feat: igraph_is_complete

aagon opened the pull request:
Follow-up of #2490.

szhorvat commented:
Let me know when this is ready for review. To fix the CodeQL CI failure, you can rebase on the latest `master` branch (but this is not critical).

szhorvat commented on interfaces/functions.yaml:
```diff
@@ -2512,6 +2512,9 @@ igraph_to_prufer:
     PARAMS: OUT GRAPH graph, INDEX_VECTOR parents, TREE_MODE type=OUT
 
+igraph_is_complete:
+    PARAMS: GRAPH graph, PRIMARY OUT BOOLEAN res
```
@ntamas What does `PRIMARY` do?

ntamas replied:
I guess it's not needed here, but most likely it doesn't hurt. According to the Stimulus docs (see `docs/index.mdx` in the Stimulus repo):

> The `PRIMARY` keyword can be used to designate one or more _output_ arguments of a function as "primary" outputs. Code generators may use this keyword to generate two versions of a function: one that returns only the primary outputs and one that returns all of them. Alternatively, code generators may also generate a _single_ function with an additional `details` argument letting the user decide whether the result should include only the primary results or all of them. Again, it is up to the code generator to support the `PRIMARY` keyword.

ntamas replied:
Curently the R code generator supports this. If there are both primary and non-primary out-args, the R code generator adds a `details=FALSE` argument to the function and only the primary out-args will be returned if `details=FALSE`.

szhorvat replied:
Ah, so like `is_tree()`. Then we indeed do not need this here.

aagon replied:
I've removed it

aagon commented:
It can be reviewed now, or I can add the support for directed graphs, or even `igraph_is_clique`, whatever you prefer.

szhorvat commented:
I had a look at the code (not the test though). Generally, it is very well written.

Regarding the ordering of neighbour lists: it is indeed guaranteed that these are sorted. This is documented for `igraph_neighbors()` and some igraph code relies on it, including the code that "simplifies" the neighbour lists.

We are missing support for directed graphs. Can you please add this?

Regarding running `igraph_is_simple()`: On a simple graph, this function needs to check each edge. It basically performs the same traversal as the following code does, iterating through all neighbour lists. Thus if we call this function, and the graph is non-simple, we end up doing twice the amount of work we'd need to. However, on simple graphs it has no disadvantage, and it is a cheap and safe way to make good use of the cache.

Thus I agree with you that this is a good way to implement `is_complete()`.

One more thing that is missing is an overflow check when computing the largest possible number of edges. What if `vcount*vcount` overflows? A quick way would be to use the macros from `math/safe_intop.h`, which even has a `IGRAPH_SAFE_N_CHOOSE_2`. However, these macros all throw an error on overflow, and this is not what we want. If the vertex count is so high that `vcount*(vcount-1)/2` is not representable, then the graph obviously can't be complete. It would be better to compare `vcount` against the largest vertex count with which a complete graph is still representable with igraph. But I'm not sure how to compute this value at compile time ...

Maybe instead of computing it, we can just assume that we're working with either standard 32-bit or 64-bit signed integers, and hard-code the limit for these two cases, similar to how it's done here:

https://github.com/igraph/igraph/blob/07d794d00d7785e348d5d8fc80a8016157391fa0/src/math/safe_intop.c#L74-L82

szhorvat commented:
So, in the undirected case, the max possible `vcount` is $2^{32} = 4294967296$ with 64-but and $2^{16} = 65536$ with 32-bit integers.

In the directed case, this will be `3037000500` and `46341`, respectively. I hope this helps.

szhorvat commented:
Finally, we still need to be smart about computing `(vcount * vcount) - vcount) / 2` in case `vcount` is too close to the above mentioned values. We could do something like:

```c
complete_ecount = vcount % 2 == 0 ? (vcount/2) * (vcount - 1) : vcount * ((vcount-1)/2)
```

ntamas commented:
> Maybe instead of computing it, we can just assume that we're working with either standard 32-bit or 64-bit signed integers, and hard-code the limit for these two cases

I agree. It's unlikely that we'll ever want to support architectures that are not 32-bit or 64-bit anyway.

szhorvat commented on src/properties/complete.c:
```diff
@@ -0,0 +1,118 @@
+ * \param graph The graph object to analyse.
+ * \param res Pointer to a logical variable, the result will be stored here.
+ *
+ * Time complexity: O(1) for simple and sparse graphs, O(|E|) at worst for graphs
+ * with either loops or multiple edges.
+ *
+ * Memory complexity: O(1) for simple and sparse graphs, O(|V|) at worst for
+ * graphs with either loops or multiple edges.
```
The time complexity would be O(|V| + |E|), i.e. no better than `igraph_is_simple()`, which this function calls. Of course if it was already cached that the graph is simple, then this is improved. But before knowing that, we can't guarantee anything better.

aagon commented:
Thank you for the review.

> Regarding the ordering of neighbour lists: it is indeed guaranteed that these are sorted.

I did not mean the neighbour list, but the list of all edges in the graph, iterated on according to edge ID. Those are not always sorted according to the "lowest adjacent vertex -- highest adjacent vertex" scheme I was mentioning, unless I'm mistaken. I asked because it would have allowed to count the edges directly on this list, rather than rely on `igraph_i_neighbours`. This would have avoided to check edges twice on undirected graphs. But this was too complicated anyway.

> We are missing support for directed graphs. Can you please add this?

Done.

> In the directed case, this will be 3037000500 and 46341, respectively. I hope this helps.

This was most useful, I am glad not to have had to do this myself. I've added a comment on how those numbers are produced.

> `complete_ecount = vcount % 2 == 0 ? (vcount/2) * (vcount - 1) : vcount * ((vcount-1)/2)`

This was very smart indeed, it is guaranteed to be very cheap.

I think that the result of this function should be cached in a cache entry "is_complete". If you agree, I will add a placeholder comment to remind myself to add caching once you're finished refactoring it.

szhorvat commented on tests/unit/igraph_is_complete.c:
```diff
@@ -0,0 +1,183 @@
+
+#include "test_utilities.h"
+
+static int check(const igraph_bool_t directed, const igraph_integer_t vertices)
```
Can you please add a _very brief_ comment about what this function does, especially starting at _ /* If directed, we must add the mutuals manually */_? I'm a bit confused about this part.

szhorvat reviewed (changes requested):
I added a few comments, but haven't had time to examine the test in detail yet.

Can you please include this function in the docs? I suggest adding it in `docs/cliques.xxml`. Feel free to propose a different location if you find one.

I'll get back to this later, including answering your questions.

szhorvat commented:
> I think that the result of this function should be cached in a cache entry "is_complete". If you agree, I will add a placeholder comment to remind myself to add caching once you're finished refactoring it.

I thought about this a bit, and it seems to me that this property would not be needed very often. If the graph is simple, caching that already achieved O(1). So caching whether the graph is complete is relevant only for multigraphs.

The original motivation for caching was to avoid running the most common checks multiple times. There are several functions which need to check if the graph is simple, or connected, before running their algorithm. Thus these checks were frequently run redundantly.

Do you have a use case in mind where it is important to not repeat complete graph checks on a multigraph?

aagon commented on src/properties/complete.c:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    /* If we arrive here, we have found no neighbour vector of size stricly less
+       than vcount - 1. The graph is therefore complete */
```
Corrected, thanks.

aagon commented on src/properties/complete.c:
```diff
@@ -0,0 +1,157 @@
+ * \param graph The graph object to analyse.
+ * \param res Pointer to a logical variable, the result will be stored here.
+ *
+ * \return Error code.
```
Done, thanks.

aagon commented on src/properties/complete.c:
```diff
@@ -0,0 +1,157 @@
+/* -*- mode: C -*-  */
+/* vim:set ts=4 sw=4 sts=4 et: */
+
```
I don't, I merely found them in other source files and copied them.

aagon commented:
> Can you please include this function in the docs? I suggest adding it in `docs/cliques.xxml`.

Done. I'm fine with adding it there, but then maybe I should change the header file in which it is declared ? I'll let you decide.

> Can you please add a very brief comment about what this function does, especially starting at _ /* If directed, we must add the mutuals manually */_? I'm a bit confused about this part.

I refactored the unit test, added comments, and used the occasion to remove all the unneeded memory management and calls to `IGRAPH_CHECK`.

> If the graph is simple, caching that already achieved O(1).

Agreed. But in that case, we should try and cache the simplicity of a graph wherever we can. For instance, a graph generated with `igraph_grg_game` is necessarily simple, but this is not cached (I thought of this because this is how I generate my graphs, but there may be a lot of other places).

> Do you have a use case in mind where it is important to not repeat complete graph checks on a multigraph?

No. I do not use multigraphs. I agree with your decision to not cache the completeness. But then, see my other point about caching simplicity much more aggressively (I can help track the places where this can be done, if you want).

szhorvat commented:
@ntamas What do you think about adding an `igraph_bool_t` directed parameter here, and making it possible to treat directed graphs as undirected?  Clique functions ignore edge directions (although they do issue a warning). Perhaps there is value in making this possible?

ntamas commented:
> What do you think about adding an igraph_bool_t directed parameter here, and making it possible to treat directed graphs as undirected?

Seems OK to me, we do this in many other places. But I th9nk we can merge the current PR as is and then we can add this later, right?

ntamas reviewed (approved):
Okay, this looks good to me then.

szhorvat commented:
@aagon Would you be interested in implementing `igraph_is_clique()` as well?
